app/EventHandlers/TrackBannerDonationRedirects.php

```diff
@@ -0,0 +1,46 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\App\EventHandlers;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\HttpKernel\Event\RequestEvent;
+use Symfony\Component\HttpKernel\KernelEvents;
+use WMDE\Fundraising\Frontend\Infrastructure\BannerSubmissionTracker;
+
+class TrackBannerDonationRedirects implements EventSubscriberInterface {
+
+	private BannerSubmissionTracker $bannerSubmissionTracker;
+	private array $submissionRoutes;
+	private string $urlParameter;
+
+	public static function getSubscribedEvents() {
+		return [
+			KernelEvents::REQUEST => 'onKernelRequest',
+		];
+	}
+
+	public function __construct( BannerSubmissionTracker $bannerSubmissionTracker, array $submissionRoutes, string $urlParameter ) {
+		$this->bannerSubmissionTracker = $bannerSubmissionTracker;
+		$this->submissionRoutes = $submissionRoutes;
+		$this->urlParameter = $urlParameter;
+	}
+
+	public function onKernelRequest( RequestEvent $event ): void {
+		$request = $event->getRequest();
+
+		if ( !in_array( $request->attributes->get( '_route' ), $this->submissionRoutes ) ) {
+			return;
+		}
+
+		if ( !$request->get( $this->urlParameter ) ) {
```

Comment:
This `if` statement will fail more often than the route check (since most banners and donors don't skip the address page). Could you put this at the top and the route check below it? That might save us some nanoseconds and micrograms of CO2.